maggy/optimizer/bayes/base.py

```diff
@@ -0,0 +1,1016 @@
+
+    Do not initialize this class! only it's subclasses
+
+    **async bo**
+
+    Bayesian Optimization consists of a surrogat model to approximate the black-box function and an acquisition
+    function to select the next hyperparameter configuration. In an asynchronous setting, we need to encourage diversity
+    when maximizing the acquisition function to prevent redundant sampling.
+
+    **optimization direction**
+
+    all optimizations are converted to minimizations problems via the `get_metrics_dict()` and `get_metrics_array()`
+    methods, where the metrics are negated in case of a maximization.
+    In the `final_store` the original metric values are saved.
+
+    **pruner**
+
+    Pruners can be run as a subroutine of an optimizer. It's interface `pruner.pruning_routine()` is called at the beginning
+    of the `get_suggestion()` method and returns the hparam config and budget for the trial. The `trial_id` of the
+    newly created Trial object is reported back to the pruner via `pruner.report_trial()`.
+
+    **budget**
+
+    If a pruner was specified ( and hence a multi-fidelity optimization is conducted ), the budget is added as an
+    additional hyperparameter to the hparam dict so it is passed to the map_fun. Note, that when fitting the
+    surrogate model the "budget" key is obviously omitted.
+
+    **models**
+
+    The surrogate models for different budgets are saved in the `models` dict with the budgets as key. In case of
+    a single fidelity optimization without a pruner or pruner with interim_results, The only model has the key `0`.
+    If we fit multiple models, sample new hparam configs always from the largest model available (i.e. biggest budget)
+
+    **imputing busy trials** (poss. shift to simple.py)
+
+    the imputed metrics are calculated on the fly in `get_metrics_array(include_busy_locations=True, budget=`budget`)`
+    for currently evaluating trials that were sampled from model with `budget`
+
+    **interim results**
+
+    In literature most surrogate models of BO algorithms are fit with the hyperparameter config and the final metric
+    of a trial.
+    Using additionally interim metrics of trials, i.e. use metrics that were trained for varying budgets makes it possible
+    to use early stopping in BO. (Also it makes it possible to have only one model in multi fidelity bo instead of one
+    model per fidelity)
+    For each interim metric a hyperparameter config augumented with the budget the metric was generated with is added
+    to the training data of the surrogate model
+
+    .. math::
+                                    z_t = [x_t, n_t] ; y_t = y_{t,nt}
+
+    When maximizing the acquisition function to sample the next hyperparameter config to evaluate, always augument with
+    the max budget N
+
+    .. math::
+                                    xt ← argmax acq([x, N])
+
+
+    """
+
+    def __init__(
+        self,
+        num_warmup_trials=15,
+        random_fraction=0.33,
```

Comment:
are these the defaults you had also in experiments?

Reply:
yes